api/models/metadata.go

```diff
@@ -0,0 +1,230 @@
+package models
+
+import (
+	"encoding/json"
+	"fmt"
+	"bytes"
+	"database/sql/driver"
+	"errors"
+)
+
+// Metadata encapsulates key-value metadata. The structure is immutable via its public API and nil-safe for its contract
+// permissive nilability is here to simplify updates and reduce the need for nil handling in extensions - metadata should be updated by over-writing the original object:
+//  oldMd := EmptyMetadata()
+//  newMd := oldMd.With("fooKey",1)
+//  // old MD remains empty
+// Metadata is lenable
+type Metadata map[string]*metadataValue
+
+// metadataValue encapsulates a metadata value in the metadata map,
+// This is stored in its un-parsed JSON format for later (re-) parsing into specific structs or values
+// metadataValue objects are  immutable after JSON load
+type metadataValue struct {
+	val []byte
+}
+
+const (
+	maxMetadataValueBytes = 512
```

Comment:
Is this a reasonable limit if the metadata structure is nested? or it it always supposed to be flat?

Reply:
it's a reasonable limit in that it keeps max metadata below 100k which was my target.